### ISSUE TITLE
Ambiguous Authentication error when access web service using proxy generated by Microsoft WCF Web Service Reference Provider  (Microsoft WCF Web Service Reference Provider Not Generating Found Services)

### ISSUE THREAD

**Issue by Lxiamail** [dotnet/wcf#3078]
@DanJ210 commented on [Tue Jul 31 2018](https://github.com/dotnet/core/issues/1829)

Hello, I'm hoping someone can shed some light into what's going on here. I'm trying to add the NetSuite Web Service (which I hate btw) and I'm using the built in feature to add a WCF service. It seems to work correctly as you'll see in the screenshot, it does find all of the operations. This is the web service URL https://webservices.netsuite.com/wsdl/v2018_1_0/netsuite.wsdl

![ns_webservice_reference](https://user-images.githubusercontent.com/16159685/43482969-35d6b74e-94d8-11e8-932a-bf4ee1eea118.PNG)

The problem is that when I click finish, the result seems to be way different than what's found. The Reference.cs that's generated contains all of these async methods and the code snippet below seems to show that it's taking the original operations and converting them to an async Task. I'm not opposed to this but the problem is that the Task that is returned is not working. It's requiring way more parameters than what the original operation requires and I'm being denied access due to "Ambiguous Authentication" which is most likely because I'm providing login details with token based authentication where I just want token. Even if I modify the generated method I'm still not getting the correct response. The original operation works.

So is there a way to not have all of this generated/converted into async methods and just get the original operations? Below is an example of what looks like the original `get` operation being converted to an `getAsync` operation. I've never worked with web services or SOAP before. I'm just trying to use .Net Core instead of .Net 4.6.

```
[System.ServiceModel.OperationContractAttribute(Action="get", ReplyAction="*")]
        [System.ServiceModel.FaultContractAttribute(typeof(ServiceNetSuite.InsufficientPermissionFault), Action="get", Name="insufficientPermissionFault", Namespace="urn:faults_2018_1.platform.webservices.netsuite.com")]
        [System.ServiceModel.FaultContractAttribute(typeof(ServiceNetSuite.InvalidSessionFault), Action="get", Name="invalidSessionFault", Namespace="urn:faults_2018_1.platform.webservices.netsuite.com")]
        [System.ServiceModel.FaultContractAttribute(typeof(ServiceNetSuite.InvalidCredentialsFault), Action="get", Name="invalidCredentialsFault", Namespace="urn:faults_2018_1.platform.webservices.netsuite.com")]
        [System.ServiceModel.FaultContractAttribute(typeof(ServiceNetSuite.ExceededConcurrentRequestLimitFault), Action="get", Name="exceededConcurrentRequestLimitFault", Namespace="urn:faults_2018_1.platform.webservices.netsuite.com")]
        [System.ServiceModel.FaultContractAttribute(typeof(ServiceNetSuite.ExceededRequestLimitFault), Action="get", Name="exceededRequestLimitFault", Namespace="urn:faults_2018_1.platform.webservices.netsuite.com")]
        [System.ServiceModel.FaultContractAttribute(typeof(ServiceNetSuite.ExceededUsageLimitFault), Action="get", Name="exceededUsageLimitFault", Namespace="urn:faults_2018_1.platform.webservices.netsuite.com")]
        [System.ServiceModel.FaultContractAttribute(typeof(ServiceNetSuite.ExceededRecordCountFault), Action="get", Name="exceededRecordCountFault", Namespace="urn:faults_2018_1.platform.webservices.netsuite.com")]
        [System.ServiceModel.FaultContractAttribute(typeof(ServiceNetSuite.ExceededRequestSizeFault), Action="get", Name="exceededRequestSizeFault", Namespace="urn:faults_2018_1.platform.webservices.netsuite.com")]
        [System.ServiceModel.FaultContractAttribute(typeof(ServiceNetSuite.UnexpectedErrorFault), Action="get", Name="unexpectedErrorFault", Namespace="urn:faults_2018_1.platform.webservices.netsuite.com")]
        [System.ServiceModel.XmlSerializerFormatAttribute(SupportFaults=true)]
        [System.ServiceModel.ServiceKnownTypeAttribute(typeof(SoapFault))]
        [System.ServiceModel.ServiceKnownTypeAttribute(typeof(BaseRef))]
        System.Threading.Tasks.Task<ServiceNetSuite.getResponse> getAsync(ServiceNetSuite.getRequest request);
```

---

@leecow commented on [Tue Jul 31 2018](https://github.com/dotnet/core/issues/1829#issuecomment-409436822)

@Lxiamail

**Comment by mlacouture** [comment]
Hi @DanJ210
The code generated by this tool implements the [Async pattern](https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/task-based-asynchronous-pattern-tap) as it is more suitable for client applications that perform async operations like calling into a web service operation.  I don't think this is your problem.

It seems that the real problem is related to the authentication method you are using as you pointed out.  The way to configure the client authentication is by implementing the partial method in the generated client class, the method signature looks like the following:

`static partial void ConfigureEndpoint(System.ServiceModel.Description.ServiceEndpoint serviceEndpoint, System.ServiceModel.Description.ClientCredentials clientCredentials);`

In this case you the interesting parameter is the `clientCredentials ` provided for this purpose.
hope this helps,

**Comment by DanJ210** [comment]
@mlacouture Hey and thanks. It looks like the generated async methods were taking all parameters that CAN be used including client credentials and token based credentials and then setting those variables to the generated `inValue` which looks to be what builds the XML request. Well the server denies the request when multiple authentication methods are provided, so I had to comment out either the client credentials being set or the TBA credentials being set. When I only used one is when I wasn't getting the ambiguous error. So no my error wasn't based on anything wrong that the generated code was creating.

I'm so new to web services and I don't like them very much but I have no choice at the moment. I love .Net Core and I love that I get generated async methods. I just have to learn more and figure out how all of this generated code is working together. I knew nothing about web services before a couple days ago and I've only been developing for a year.

Too much info but thanks very much for the pointer above. It does help and I will look into what you suggested.

Thanks very much.
